modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4IncrementalRequestHandlingIT.java

```diff
@@ -0,0 +1,429 @@
+                        if (recvChunk.isLast) {
+                            break;
+                        }
+                    }
+                }
+
+                assertEquals("sent and received payloads are not the same", sendData, recvData);
+                handler.sendResponse(new RestResponse(RestStatus.OK, ""));
+            }
+            assertBusy(() -> assertEquals("should receive all server responses", totalRequests, ctx.clientRespQueue.size()));
+        }
+    }
+
+    // ensures that all queued chunks are released when connection closed
+    public void testClientConnectionCloseMidStream() throws Exception {
+        try (var ctx = setupClientCtx()) {
+            var opaqueId = opaqueId(0);
+
+            // write half of http request
+            ctx.clientChannel.write(httpRequest(opaqueId, 2 * 1024));
+            ctx.clientChannel.writeAndFlush(randomContent(1024, false));
+
+            // await stream handler is ready and request full content
+            var handler = ctx.awaitRestChannelAccepted(opaqueId);
+            assertBusy(() -> assertEquals(1, handler.stream.chunkQueue().size()));
+
+            // enable auto-read to receive channel close event
+            handler.stream.channel().config().setAutoRead(true);
+
+            // terminate connection and wait resources are released
+            ctx.clientChannel.close();
+            assertBusy(() -> assertEquals(0, handler.stream.chunkQueue().size()));
+        }
+    }
+
+    // ensures that all queued chunks are released when server decides to close connection
+    public void testServerCloseConnectionMidStream() throws Exception {
+        try (var ctx = setupClientCtx()) {
+            var opaqueId = opaqueId(0);
+
+            // write half of http request
+            ctx.clientChannel.write(httpRequest(opaqueId, 2 * 1024));
+            ctx.clientChannel.writeAndFlush(randomContent(1024, false));
+
+            // await stream handler is ready and request full content
+            var handler = ctx.awaitRestChannelAccepted(opaqueId);
+            assertBusy(() -> assertEquals(1, handler.stream.chunkQueue().size()));
+
+            // terminate connection on server and wait resources are released
+            handler.channel.request().getHttpChannel().close();
+            assertBusy(() -> assertEquals(0, handler.stream.chunkQueue().size()));
+        }
+    }
+
+    // ensure that client's socket buffers data when server is not consuming data
+    public void testClientBackpressure() throws Exception {
+        try (var ctx = setupClientCtx()) {
+            var opaqueId = opaqueId(0);
+            var payloadSize = MBytes(50);
+            ctx.clientChannel.writeAndFlush(httpRequest(opaqueId, payloadSize));
+            for (int i = 0; i < 5; i++) {
+                ctx.clientChannel.writeAndFlush(randomContent(MBytes(10), false));
+            }
+            ctx.clientChannel.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);
```

Comment:
Can we assert that this future's `isDone()` returns false because of the backpressure?